framework/src/main/java/org/checkerframework/common/basetype/BaseTypeChecker.java

```diff
@@ -497,14 +505,99 @@ public void typeProcess(TypeElement element, TreePath tree) {
+        this.elementsWithSuppressedWarnings.clear();
+        Set<String> checkerKeys = new HashSet<>(getSuppressWarningsKeys());
+        Set<String> errorKeys = new HashSet<>(messages.stringPropertyNames());
+        for (BaseTypeChecker subChecker : subcheckers) {
+            elementsSuppress.addAll(subChecker.elementsWithSuppressedWarnings);
+            subChecker.elementsWithSuppressedWarnings.clear();
+            checkerKeys.addAll(subChecker.getSuppressWarningsKeys());
+            errorKeys.addAll(subChecker.messages.stringPropertyNames());
+            subChecker.getVisitor().treesWithSuppressWarnings.clear();
+        }
+        // It's not clear for which checker this suppression is intended,
+        // so never report it as unused.
+        checkerKeys.remove(SourceChecker.SUPPRESS_ALL_KEY);
+
+        for (Tree tree : getVisitor().treesWithSuppressWarnings) {
+            Element elt = TreeUtils.elementFromTree(tree);
+            SuppressWarnings suppressAnno = elt.getAnnotation(SuppressWarnings.class);
+            if (suppressAnno == null || elementsSuppress.contains(elt)) {
+                continue;
+            }
+            for (String keyFromAnno : suppressAnno.value()) {
+                for (String checkerKey : checkerKeys) {
+                    if (keyFromAnno.contains(checkerKey)) {
+                        Tree swTree = findSWTree(tree);
+                        report(
+                                Result.warning(
+                                        SourceChecker.UNNEEDED_SUPPRESSION_KEY,
+                                        getClass().getSimpleName(),
+                                        "\"" + keyFromAnno + "\""),
+                                swTree);
+                    }
+                }
+                if (keyFromAnno.contains(":")) {
+                    // The key starts with a checker name, if that is this checker, then the warning
+                    // was issued above.  For example, if this is the Nullness Checker and the
+                    // keyForAnno is "index:override.return.invalid", then don't issue a warning.
+                    continue;
+                }
+
+                for (String errorKey : errorKeys) {
+                    if (keyFromAnno.equals(errorKey)) {
+                        Tree swTree = findSWTree(tree);
+                        report(
+                                Result.warning(
+                                        SourceChecker.UNNEEDED_SUPPRESSION_KEY,
+                                        getClass().getSimpleName(),
+                                        // Have to add quotes here so that the key not its value is printed.
+                                        "\"" + keyFromAnno + "\""),
+                                swTree);
+                    }
+                }
+            }
+        }
+        getVisitor().treesWithSuppressWarnings.clear();
+    }
+
+    private Tree findSWTree(Tree tree) {
```

Comment:
Some documentation would be helpful to understand what this is doing. Sometime it returns annotation trees, sometime just the input parameter.
Wouldn't it be nicer to use a more precise return type and raise an error if the unexpected `return tree` happens?
But maybe I misunderstand the purpose.